engine/src/main/java/org/terasology/rendering/world/WorldRendererImpl.java

```diff
@@ -195,13 +199,21 @@ private void initRenderingSupport() {
         shaderManager.initShaders();
         postProcessor.initializeMaterials();
         initMaterials();
+
+        shadowMap = new ShadowMapNode(getMaterial("engine:prog.shadowMap"),
+                playerCamera,
+                this,
+                backdropProvider,
+                renderingConfig,
+                graphicState);
+
```

Comment:
Generally speaking a RenderNode should somehow find everything it needs. This way the DAG remains agnostic of what a node does or needs and can treat all nodes the same.

So:
1. let the node find the renderingConfig and the backdropProvider via a context object and let it obtain the playerCamera from the WorldRenderer instance.

2. GraphicState is a class I created to collect easily recognizable opengl directives in one place. Earlier they were "visually polluting" the renderer code making it hard to get an overview - like a picture with too many details and no focus to guide the eye. However, as we divide the code of the existing rendering engine into RenderNodes, it will make sense to move those gl directives into the nodes, near where it's needed. The code within a node will never be horribly big, probably. So, what you could do is to place the code currently in GraphicState.preRenderSetupSceneShadowMap() and GraphicState.postRenderCleanupSceneShadowMap() within the ShadowMapNode, eliminating the dependency on the graphicState instance.

Word of caution though: I remind you that at this stage generating the ShadowMap or executing any other step in the rendering engine might still rely on state changes occurred much earlier in the flow of execution. Until every state change is first done and then undone immediately after use, leaving to the DAG to optimize the process, you must ensure that RenderNodes are executed in the current order. Let me know if you need me to further clarify this.